feat: replace redirect url pattern with route key pattern

zhongliang02 opened the pull request:
## Context

Due to yet another `open redirect` VAPT finding (`care360: GTA-55-008`), it has proved to be very challenging to validate the `callbackUrl` parameter.

## Approach

This PR completely removes the `callbackUrl` parameter in favor of the `redirectRouteKey` parameter.

In all situations where we redirect ,`redirectRouteKey` is first validated to be `keyof AllRoutes`, falling back to `HOME` when invalid. (`AllRoutes` is a dict with all the routes in the app).

Then we resolve the `route key` to the actual route.

## How to read this PR

Please look at all the utilities `lib/` `schemas/` `utils/` first to familiarize with the constructs that enable this route key pattern.

Then evaluate the components involved in the email login flow followed by the SGID login flow.

Finally you can evaluate utility component wrappers under `AuthWrappers/`

## Risks and Testing

I will manually test the vercel deployment to ensure that both login flows still work as intended.

karrui commented on src/lib/routes.ts:
```diff
@@ -1,6 +1,14 @@
 export const PROFILE = '/profile'
 export const SETTINGS_PROFILE = '/settings/profile'
+
+export const AllRoutes = {
```
nit: since this is a constant, should be in UPPER_CAMEL_CASE

```suggestion
export const ALL_ROUTES = {
```

karrui commented on src/components/AuthWrappers/EnforceLoginStatePageWrapper.tsx:
i don't think the API has to change much here right? seems like we might only need to update callbackUrlSchema to allow specific `redirectUrl`s declared by the user (and fallback to a default if the parse fails). Everything should stay the same?

karrui commented on src/components/AuthWrappers/EnforceLoginStatePageWrapper.tsx:
```diff
@@ -1,45 +1,39 @@
-  void router.replace(
-    callbackUrlSchema.parse(appendWithRedirect(redirectTo, redirectUrl)),
-  )
+  useEffect(() => {
```
comment: was there a bug not using `useEffect` previously? actually yeah might be better to use an effect. Maybe this was causing the infinite loading spinner that may sometimes trigger :O

zhongliang02 replied:
I think many properties of `router` are undefined on first render, so I always have a habit of putting router related things in a `useEffect`

karrui commented on src/components/AuthWrappers/PublicPageWrapper.tsx:
```diff
@@ -24,12 +24,10 @@ export const PublicPageWrapper = ({
   const { hasLoginStateFlag } = useLoginState()
 
   if (hasLoginStateFlag && rest.strict) {
```
should this be in a `useEffect` too?

zhongliang02 replied:
done

karrui commented on src/features/sign-in/components/SgidCallback.tsx:
```diff
@@ -19,27 +20,29 @@ export const SgidCallback = (): JSX.Element => {
+  )
 
   useEffect(() => {
-    if (!selectProfileStep) {
-      setHasLoginStateFlag()
+    if (!response.success) {
+      void router.replace(
```
should this be run through the zod validation schema too?

zhongliang02 replied:
I think it's not necessary because SIGN_IN is a hardcoded constant

karrui commented on src/features/sign-in/components/SgidLogin/SgidLoginButton.tsx:
```diff
@@ -3,23 +3,20 @@ import { Box, Divider, Flex, HStack, Stack, Text } from '@chakra-ui/react'
   const sgidLoginMutation = trpc.auth.sgid.login.useMutation({
     onSuccess: async ({ redirectUrl }) => {
-      await router.push(redirectUrl)
+      window.location.href = redirectUrl
```
when using nextjs, it is best practice to check for the location of `window.location` first, since this component might be SSR'd.
this might be a subtle bug if somehow invoked from the server. 

why was the `router.push` call replaced with this? was there a particular reason?

zhongliang02 replied:
hmm it's because redirectUrl is already known to be an external URL, so I thought there'd be no benefits from `router.push`ing it, but I was not aware of the possibility of SSR'ing the component, I thought it's always CSR unless there's `getServerSideProps`.

karrui commented on src/utils/url.ts:
```diff
@@ -1,28 +1,44 @@
+export const resolveRouteKey = (
+  v: unknown,
+): (typeof AllRoutes)[keyof typeof AllRoutes] => {
+  return AllRoutes[routeKeySchema.parse(v)]
```
should we return a fallback here? unsure, but better than users not changing routes maybe?

zhongliang02 replied:
`routeKeySchema.parse(v)` will fallback to `HOME` after parsing. `routeKeySchema` is guaranteed to produce `keyof AllRoutes`

karrui reviewed (commented):
i feel like we can keep the changes to a minimum, instead of changing all the variables to something new. The concept is still the same right, just that the "redirect URL"s are now restricted to what the user has explicitly set. should we just change the callbackUrl schema and keep everything else the same? 

(trying to make it easy for starter-kit derived project to update their code too, since there might be too many things to change in here for them)

zhongliang02 commented:
Re: minimizing changes to variables, I think conceptually it is quite different though,

previous: `?callbackUrl=/home`
new: `?redirectRouteKey=HOME`, then `ALL_ROUTES = {'HOME': '/home'}` and we just `router.push(ALL_ROUTES[...])`

In theory we can remove this layer of indirection and change it to something like
`?callbackUrl=/home` and `ALLOWED_ROUTES = ['/home', ...]`

I'll make a commit with the above and we can see how that looks

zhongliang02 commented:
Reverted to the redirect url pattern, but now `callbackUrlSchema` simply follows a whitelist. 
Technically this is more of a kitchen sink PR now, lmk if you prefer if I split it up to
- callbackUrlSchema changes
- useEffect fixes
- sgid login minor refactoring

karrui commented on src/components/AuthWrappers/EnforceLoginStatePageWrapper.tsx:
```diff
@@ -1,45 +1,37 @@
 /**
  * Page wrapper that renders children only if the login state localStorage flag has been set.
- * Otherwise, will redirect to the route passed into the `redirectTo` prop.
+ * Otherwise, will redirect to SIGN_IN, which will redirect to route passed into the `redirectTo` prop.
```
seems like this will redirect to `HOME` instead? is that intended?

zhongliang02 replied:
The component redirects to `/sign-in?callbackUrl=<redirectTo>` when not signed in, `redirectTo` defaults to `HOME` if not provided.

karrui commented on src/lib/routes.ts:
```diff
@@ -1,6 +1,16 @@
 export const PROFILE = '/profile'
 export const SETTINGS_PROFILE = '/settings/profile'
+
+export const ALLOWED_CALLBACK_ROUTES = [
```
i wonder if this would work for dynamic routes (like `/user/[userId]`). If not, might break many autoredirects/ cause confusion.

zhongliang02 replied:
No it doesnt, it will definitely break dynamic routes.

The issue is because `router.push` does too much magic (e.g. you can do `router.push('/home/[x]?x=../signout')` to `/signout`)

It became quite challenging to validate that `router.push` does not cause unintended navigation with user supplied input (`?callbackUrl` param) (especially across next versions), so the alternative is to use a hardcoded whitelist approach like the PR shows.

`appUrlSchema` can still be used to validate urls if the dev wishes to use dynamic urls, just that the additional friction is supposed to discourage that

zhongliang02 replied:
Happy to hear about alternatives though, not sure what's the best approach tbh
